### PR TITLE
:rocket: Feature/fit model

### PR DESCRIPTION
**Pull request by raodz**
## Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Tests update
- [ ] Documentation update
- [ ] Changes that do not directly affect the source code or tests

## Checklist
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/dummy_models.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    return np.ones(len(val_data), dtype=int)
+
+
+def dummy_always_non_bald(val_data):
+    return np.zeros(len(val_data), dtype=int)
+
+
+def dummy_random(val_data):
+    return np.random.randint(2, size=len(val_data))
+
```
Comment (jakub1090cn) [inline src/dummy_models.py:17]: I would add something like DummyClassifier(strategy="most_frequent") from sklearn.dummy

File: scripts/run_dummy_models_on_val.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import pandas as pd
+import os
+from src.dummy_models import (
+    dummy_random,
+    dummy_always_bald,
+    dummy_always_non_bald,
+    evaluate_dummy_model,
+)
+
```
Comment (skrzypczykt) [inline scripts/run_dummy_models_on_val.py:3]: main

File: scripts/run_dummy_models_on_val.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+val_csv_path = os.path.join("..", "src", "data", "val.csv")
```
Comment (skrzypczykt) [inline scripts/run_dummy_models_on_val.py]: use dummy_evaluatin_config

File: scripts/run_dummy_models_on_val.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+val_df = pd.read_csv(val_csv_path)
+val_labels = val_df["label"].values
+
+output_dir = os.path.join("..", "results")
+os.makedirs(output_dir, exist_ok=True)
+output_file = os.path.join(output_dir, "dummy_model_on_val_metrics_report.txt")
```
Comment (skrzypczykt) [inline scripts/run_dummy_models_on_val.py]: dummy_model_on_val_metrics_report.txt to consts

File: scripts/run_dummy_models_on_val.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+with open(output_file, "w") as report_file:
```
Comment (skrzypczykt) [inline scripts/run_dummy_models_on_val.py]: first calculate what you nedd, than open a file and write.

Extract saving the results to  function from utils and use here and in training pipeline

File: scripts/run_dummy_models_on_val.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import pandas as pd
+import os
+from src.dummy_models import (
+    dummy_random,
```
Comment (skrzypczykt) [inline scripts/run_dummy_models_on_val.py]: rename to dummy_random**_model** or predictor

File: src/dummy_models.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    precision = precision_score(true_labels, predictions)
+    recall = recall_score(true_labels, predictions)
+    f1 = f1_score(true_labels, predictions)
+
+    return {
+        "accuracy": accuracy,
+        "precision": precision,
+        "recall": recall,
+        "f1_score": f1,
+    }
```
Comment (skrzypczykt) [inline src/dummy_models.py]: do not reinvent the wheel

File: scripts/tune_hyperparameters.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+from src.config_class import BaldOrNotConfig
+from src.constants import N_CHANNELS_RGB, DEFAULT_IMG_SIZE
+
+
+def main():
+    # Load the configuration
+    config = BaldOrNotConfig()
+
+    # Initialize the datasets
+    train_csv_path = os.path.join("..", "src", "data", "train.csv")
```
Comment (skrzypczykt) [inline scripts/tune_hyperparameters.py]: to config

File: scripts/tune_hyperparameters.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    train_dataset = BaldDataset(
+        train_df,
+        batch_size=config.tuning_params.batch_size,
+        dim=DEFAULT_IMG_SIZE,
+        n_channels=N_CHANNELS_RGB,
+        shuffle=True,
+        augment_minority_class=config.tuning_params.augment_minority_class,
+    )
+
+    val_csv_path = os.path.join("..", "src", "data", "val.csv")
```
Comment (skrzypczykt) [inline scripts/tune_hyperparameters.py]: to config

File: src/tuning.py [hunk 10]
```diff
@@ -1,169 +1,75 @@
...
-    }
-
-    param_combinations = list(
-        product(
-            batch_sizes,
-            max_class_imbalance_ratios,
-            use_class_weight_options,
-            augment_class_options,
-        )
+def tune_model(train_dataset, val_dataset, config: BaldOrNotConfig):
```
Comment (skrzypczykt) [inline src/tuning.py:51]: use ony tuning_params not config

File: src/config_class.py [hunk 11]
```diff
@@ -21,11 +21,24 @@ class TrainingParams:
...
-    max_tuning_trials: int = 15
+    use_tuned_hyperparameters: bool = True
+
+
+@dataclass
+class TuningParams:
+    max_tuning_trials: int = 12
+    loss_function: str = "binary_crossentropy"
+    objective: str = "val_f1_score"
+    factor: int = 4
```
Comment (skrzypczykt) [inline src/config_class.py]: describe

File: src/tuning.py [hunk 12]
```diff
@@ -1,169 +1,75 @@
...
+    Builds the model for hyperparameter tuning.
+
+    Args:
+        hp: Hyperparameter object for Keras Tuner.
+        config: Configuration object containing the model parameters.
+
+    Returns:
+        A compiled Keras model.
+    """
+    hp_dense_units = hp.Choice("dense_units", values=[128, 256, 512])
```
Comment (skrzypczykt) [inline src/tuning.py]: move to config

File: src/config_class.py [hunk 13]
```diff
@@ -58,6 +71,7 @@ class BaldOrNotConfig:
     training_params: TrainingParams = field(
         default_factory=lambda: TrainingParams()
     )
+    tuning_params: TuningParams = field(default_factory=lambda: TuningParams())
```
Comment (skrzypczykt) [inline src/config_class.py:89]: use abstract config and TrainingConfig and TuningConfig

File: src/data.py [hunk 14]
```diff
@@ -429,3 +432,15 @@ def adjust_class_distribution(
...
+    @staticmethod
+    def get_classes_weights(df: pd.DataFrame):
+        n_total = len(df)
+        n_not_bald = df["label"].value_counts()[NOT_BALD_LABEL]
+        n_bald = df["label"].value_counts()[BALD_LABEL]
+        not_bald_weight = (1 / n_not_bald) * (n_total / NUMBER_OF_CLASSES)
+        bald_weight = (1 / n_bald) * (n_total / NUMBER_OF_CLASSES)
+        return {
+            str(NOT_BALD_LABEL): not_bald_weight,
+            str(BALD_LABEL): bald_weight,
```
Comment (skrzypczykt) [inline src/data.py]: rm

File: src/model_training.py [hunk 15]
```diff
@@ -10,97 +11,64 @@
...
+        batch_size=config.training_params.batch_size,
         dim=DEFAULT_IMG_SIZE,
         n_channels=N_CHANNELS_RGB,
         shuffle=True,
-        augment_minority_class=augment_class,
+        augment_minority_class=config.training_params.augment_minority_class,
     )
-    logging.info(f"Training dataset initialized with batch size {batch_size}")
 
+    val_csv_path = os.path.join("..", "src", "data", "val.csv")
```
Comment (skrzypczykt) [inline src/model_training.py]: config

File: src/tuning.py [hunk 16]
```diff
@@ -1,169 +1,75 @@
...
+    Args:
+        train_dataset: Training dataset.
+        val_dataset: Validation dataset.
+        config (BaldOrNotConfig): Configuration object containing the training parameters.
+
+    Returns:
+        The best hyperparameters found during tuning.
+    """
+    tuner = kt.Hyperband(
+        lambda hp: model_builder(hp, config),
```
Comment (skrzypczykt) [inline src/tuning.py]: partial(model_builder, config=config)

File: src/tuning.py [hunk 17]
```diff
@@ -1,169 +1,75 @@
...
-            }
-            logging.info(f"New best parameters found: {best_params}")
-            print(f"New best parameters found: {best_params}")
-
-        logging.info(f"Best val_loss to this point: {best_val_loss}")
-        logging.info(f"Best params to this point: {best_params}")
-        print(f"Best val_loss to this point: {best_val_loss}")
-        print(f"Best params to this point: {best_params}")
+    # Retrieve the best hyperparameters
+    best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]
```
Comment (skrzypczykt) [inline src/tuning.py:82]: Check if you can use`*`

File: scripts/prepare_datasets.py [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    original_label=ORIGINAL_NOT_BALD_LABEL,
+    new_label=NOT_BALD_LABEL,
+)
+
+train_df, val_df, test_df = BaldDataset.create_subset_dfs(corrected_labels_df)
+
+
+data_dir = os.path.join("..", "src", "data")
+if not os.path.exists(data_dir):
+    os.makedirs(data_dir)
```
Comment (jakub1090cn) [inline scripts/prepare_datasets.py]: os.makedirs(data_dir, exist_ok=True)

File: src/model_training.py [hunk 19]
```diff
@@ -1,83 +1,78 @@
...
+    if config.training_params.tune_hyperparameters:
+        logging.info("Tuning hyperparameters...")
+        best_hps = tune_model(train_dataset, val_dataset, config)
+        config.model_params.dense_units = best_hps.get("dense_units")
+        config.model_params.dropout_rate = best_hps.get("dropout_rate")
+        config.training_params.learning_rate = best_hps.get("learning_rate")
+
+    print(config.model_params.dense_units)
+    print(config.model_params.dropout_rate)
+    print(config.training_params.learning_rate)
```
Comment (jakub1090cn) [inline src/model_training.py]: why print instead of logging?

File: src/dummy_models.py (hunk 1 above)
Comment (raodz) [reply src/dummy_models.py:17]: changed to subclass of keras Model

File: scripts/run_dummy_models_on_val.py (hunk 2 above)
Comment (raodz) [reply scripts/run_dummy_models_on_val.py:3]: ok

File: scripts/run_dummy_models_on_val.py (hunk 3 above)
Comment (raodz) [reply scripts/run_dummy_models_on_val.py]: changed to get_metrics() from evaluation.py instead

File: scripts/run_dummy_models_on_val.py (hunk 4 above)
Comment (raodz) [reply scripts/run_dummy_models_on_val.py]: ok

File: scripts/run_dummy_models_on_val.py (hunk 6 above)
Comment (raodz) [reply scripts/run_dummy_models_on_val.py]: ok

File: src/dummy_models.py (hunk 7 above)
Comment (raodz) [reply src/dummy_models.py]: ok

File: scripts/tune_hyperparameters.py (hunk 8 above)
Comment (raodz) [reply scripts/tune_hyperparameters.py]: ok

File: scripts/tune_hyperparameters.py (hunk 9 above)
Comment (raodz) [reply scripts/tune_hyperparameters.py]: ok

File: src/tuning.py (hunk 10 above)
Comment (raodz) [reply src/tuning.py:51]: ok

File: src/config_class.py (hunk 11 above)
Comment (raodz) [reply src/config_class.py]: ok

File: src/tuning.py (hunk 12 above)
Comment (raodz) [reply src/tuning.py]: ok

File: src/config_class.py (hunk 13 above)
Comment (raodz) [reply src/config_class.py:89]: ok

File: src/data.py (hunk 14 above)
Comment (raodz) [reply src/data.py]: ok

File: src/model_training.py (hunk 15 above)
Comment (raodz) [reply src/model_training.py]: ok

File: src/tuning.py (hunk 16 above)
Comment (raodz) [reply src/tuning.py]: ok

File: scripts/prepare_datasets.py (hunk 18 above)
Comment (raodz) [reply scripts/prepare_datasets.py]: ok

File: src/model_training.py (hunk 19 above)
Comment (raodz) [reply src/model_training.py]: changed

File: scripts/prepare_datasets.py [hunk 20]
```diff
@@ -0,0 +1,34 @@
+import os
+
+import pandas as pd
+
+from src.config_class import BaldOrNotConfig
+from src.constants import NOT_BALD_LABEL, ORIGINAL_NOT_BALD_LABEL
+from src.data import BaldDataset
+
```
Comment (skrzypczykt) [inline scripts/prepare_datasets.py:6]: main

File: src/config_class.py [hunk 21]
```diff
@@ -1,44 +1,52 @@
...
-    epochs: int = 200
-    batch_size: int = 64
-    learning_rate: float = 0.001
-    optimizer: str = "adam"
+class LearningParams:
+    epochs: int = 3
+    batch_size: int = 10
     loss_function: str = "binary_crossentropy"
+    max_class_imbalance_ratio: int = 1
+    steps_per_epoch: Optional[int] = 10
```
Comment (skrzypczykt) [inline src/config_class.py]: int | None

File: src/config_class.py [hunk 22]
```diff
@@ -1,44 +1,52 @@
...
-    batch_size: int = 64
-    augment_minority_class: bool = False
-    epochs: int = 5
-    steps_per_epoch: int = 30
-    validation_steps: int = 30
+class TuningParams(LearningParams):
+    max_tuning_trials: int = 3
+    objective: str = "val_f1_score"  # The metric to be optimized during tuning
+    factor: int = 3
+    hp_dense_units_values: List[int] = field(
```
Comment (skrzypczykt) [inline src/config_class.py]: list[int]

File: src/constants.py [hunk 23]
```diff
@@ -8,3 +8,4 @@
 N_CHANNELS_GRAYSCALE = 1
 DEFAULT_IMG_SIZE = (224, 224)
 LOG_FILE_NAME = "training.log"
+DUMMY_METRICS_FILE_NAME_PREFIX = "dummy_model_val_"
```
Comment (skrzypczykt) [inline src/constants.py:11]: delete suffix

File: src/model_training.py [hunk 24]
```diff
@@ -53,18 +51,9 @@ def train_model(config: BaldOrNotConfig, output_dir_path: str):
...
-        config.model_params.dense_units = best_hps.get("dense_units")
-        config.model_params.dropout_rate = best_hps.get("dropout_rate")
-        config.training_params.learning_rate = best_hps.get("learning_rate")
-
-    print(config.model_params.dense_units)
-    print(config.model_params.dropout_rate)
-    print(config.training_params.learning_rate)
-    # Build and compile the model with tuned or predefined hyperparameters
+    if config.training_params.use_tuned_hyperparameters:
+        pass
```
Comment (skrzypczykt) [inline src/model_training.py]: rm

File: src/tuning.py (hunk 17 above)
Comment (skrzypczykt) [reply src/tuning.py:82]: +1

File: src/utils.py [hunk 25]
```diff
@@ -26,3 +27,37 @@ def wrapper(*args, **kwargs):
         return func(*args, **kwargs)
 
     return wrapper
+
+
+import os
+import numpy as np
+
```
Comment (skrzypczykt) [inline src/utils.py]: move

**Review by skrzypczykt — APPROVED** [review]
cool

File: scripts/run_dummy_models_on_val.py (hunk 5 above)
Comment (raodz) [reply scripts/run_dummy_models_on_val.py]: ok

File: src/tuning.py (hunk 17 above)
Comment (raodz) [reply src/tuning.py:82]: "Can't use starred expression here"

File: scripts/prepare_datasets.py (hunk 20 above)
Comment (raodz) [reply scripts/prepare_datasets.py:6]: ok

File: src/config_class.py (hunk 21 above)
Comment (raodz) [reply src/config_class.py]: ok

File: src/config_class.py (hunk 22 above)
Comment (raodz) [reply src/config_class.py]: ok

File: src/constants.py (hunk 23 above)
Comment (raodz) [reply src/constants.py:11]: ok

File: src/model_training.py (hunk 24 above)
Comment (raodz) [reply src/model_training.py]: ok

File: src/utils.py (hunk 25 above)
Comment (raodz) [reply src/utils.py]: ok

**Review by jakub1090cn — APPROVED** [review]
cool

**Review by skrzypczykt — APPROVED** [review]
Cool, one minor thing: use any consistently. You probably don't need Any anymore.

**Pull request merged**